src/no_std_error.rs

```diff
@@ -0,0 +1,199 @@
+    ///         }
+    ///         _ => println!("No error"),
+    ///     }
+    /// }
+    /// ```
+    fn cause(&self) -> Option<&dyn Error> {
+        self.source()
+    }
+
+    /// The lower-level source of this error, if any.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use snafu::Error;
+    /// use core::fmt;
+    ///
+    /// #[derive(Debug)]
+    /// struct SuperError {
+    ///     side: SuperErrorSideKick,
+    /// }
+    ///
+    /// impl fmt::Display for SuperError {
+    ///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    ///         write!(f, "SuperError is here!")
+    ///     }
+    /// }
+    ///
+    /// impl Error for SuperError {
+    ///     fn description(&self) -> &str {
+    ///         "I'm the superhero of errors"
+    ///     }
+    ///
+    ///     fn source(&self) -> Option<&(dyn Error + 'static)> {
+    ///         Some(&self.side)
+    ///     }
+    /// }
+    ///
+    /// #[derive(Debug)]
+    /// struct SuperErrorSideKick;
+    ///
+    /// impl fmt::Display for SuperErrorSideKick {
+    ///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    ///         write!(f, "SuperErrorSideKick is here!")
+    ///     }
+    /// }
+    ///
+    /// impl Error for SuperErrorSideKick {
+    ///     fn description(&self) -> &str {
+    ///         "I'm SuperError side kick"
+    ///     }
+    /// }
+    ///
+    /// fn get_super_error() -> Result<(), SuperError> {
+    ///     Err(SuperError { side: SuperErrorSideKick })
+    /// }
+    ///
+    /// fn main() {
+    ///     match get_super_error() {
+    ///         Err(e) => {
+    ///             println!("Error: {}", e.description());
+    ///             println!("Caused by: {}", e.source().unwrap());
+    ///         }
+    ///         _ => println!("No error"),
+    ///     }
+    /// }
+    /// ```
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        None
+    }
+}
+
+impl Error for core::str::ParseBoolError {
+    fn description(&self) -> &str {
+        "failed to parse bool"
+    }
+}
+impl Error for core::str::Utf8Error {
+    fn description(&self) -> &str {
+        "invalid utf-8: corrupt contents"
+    }
+}
+impl Error for core::num::ParseIntError {}
+impl Error for core::num::TryFromIntError {}
+impl Error for core::array::TryFromSliceError {}
+impl Error for core::num::ParseFloatError {}
+impl Error for core::fmt::Error {
+    fn description(&self) -> &str {
+        "an error occurred when formatting an argument"
+    }
+}
+impl Error for core::cell::BorrowError {
+    fn description(&self) -> &str {
+        "already mutably borrowed"
+    }
+}
+impl Error for core::cell::BorrowMutError {
+    fn description(&self) -> &str {
+        "already borrowed"
+    }
+}
+impl Error for core::char::CharTryFromError {
+    fn description(&self) -> &str {
+        "converted integer out of range for `char`"
+    }
+}
+impl Error for core::char::ParseCharError {}
```

Comment:
Speaking of "tracking changes", what would we do when (for example) Rust 1.40 adds a new `core::foo::FooError`? That type won't exist in Rust 1.39, so we can't add it here without breaking backwards compatibility. If we don't add it, then people won't be able to use the fancy features, as you mentioned.

Reply:
That is actually already kind of the case with some of the Errors here, they aren't stable in 1.18. So maybe we can do rustc version detection via a build.rs? That would get rid of the 1.30 feature too.